File: engine_glue/src/sys_c/window.cpp
#include <sys_c/window.h>
#include <sys/window.h>

C_API_FUNC(SYS_C_API, uint8_t, window_closing, Window* w)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    return static_cast<uint8_t>(win->Closing());
}

C_API_FUNC(SYS_C_API, void, window_warp, Window* w, int x, int y)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->Warp(x, y);
}

C_API_FUNC(SYS_C_API, void, window_warp_relative, Window* w, int x, int y)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->WarpRelative(x, y);
}

C_API_FUNC(SYS_C_API, void, window_resize, Window* win, int w, int h)
{
    auto window = reinterpret_cast<lumi::sys::Window*>(win);
    window->Resize(w, h);
}

C_API_FUNC(SYS_C_API, void, window_set_name, Window* w, char* name)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetName(std::string(name));
}

C_API_FUNC(SYS_C_API, void, window_set_icon, Window* w, char* icon)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetIcon(std::string(icon));
}

C_API_FUNC(SYS_C_API, void, window_set_size_limits, Window* w, 
    int wMin, int wMax, 
    int hMin, int hMax)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetSizeLimits(wMin, wMax, hMin, hMax);
}

C_API_FUNC(SYS_C_API, void, window_set_width_limits, Window* w, 
    int wMin, int wMax)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetWidthLimits(wMin, wMax);
}

C_API_FUNC(SYS_C_API, void, window_set_height_limits, Window* w,
    int hMin, int hMax)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetHeightLimits(hMin, hMax);
}

C_API_FUNC(SYS_C_API, void, window_set_resizable, Window* w, 
    uint8_t resizable)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetResizable(static_cast<bool>(resizable));
}

C_API_FUNC(SYS_C_API, void, window_set_bordered, Window* w,
    uint8_t bordered)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetBordered(static_cast<bool>(bordered));
}

C_API_FUNC(SYS_C_API, void, window_set_mode, Window* w,
    int mode)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    win->SetMode(static_cast<lumi::sys::WindowMode>(mode));
}

C_API_FUNC(SYS_C_API, void, window_destroy, Window* w)
{
    delete reinterpret_cast<lumi::sys::Window*>(w);
}

C_API_FUNC(SYS_C_API, uint8_t, window_is_valid, Window* w)
{
    auto win = reinterpret_cast<lumi::sys::Window*>(w);
    return static_cast<uint8_t>(win->IsWindow());
}

File: EngineManaged/Libraries/Lumi.Sys/Window.cs
namespace Lumi.Sys;

public class Window
{
    private IntPtr _nativePtr;
    public static int GetWindowPosCentered() => 0x2FFF0000;

    internal Window(IntPtr nativeHandle) => _nativePtr = nativeHandle;
    public void Warp(int x, int y) => WindowNative.Warp(_nativePtr, x, y);
    public void WarpRelative(int x, int y) => WindowNative.WarpRelative(_nativePtr, x, y);
    public void Resize(int w, int h) => WindowNative.Resize(_nativePtr, w, h);
    public void SetName(string name) => WindowNative.SetName(_nativePtr, name);
    public void SetIcon(string icon) => WindowNative.SetIcon(_nativePtr, icon);
    public void SetSizeLimits(int wMin, int wMax, int hMin, int hMax)
        => WindowNative.SetSizeLimits(_nativePtr, wMin, wMax, hMin, hMax);

    public void SetWidthLimits(int wMin, int wMax)
        => WindowNative.SetWidthLimits(_nativePtr, wMin, wMax);

    public void SetHeightLimits(int hMin, int hMax)
        => WindowNative.SetHeightLimits(_nativePtr, hMin, hMax);

    public void SetResizable(bool resizable)
        => WindowNative.SetResizable(_nativePtr, resizable);

    public void SetBordered(bool bordered)
        => WindowNative.SetBordered(_nativePtr, bordered);

    public void SetMode(WindowMode mode) => WindowNative.SetMode(_nativePtr, mode);

    public bool Closing() => WindowNative.Closing(_nativePtr);
    public void Delete()
    {
        if (_nativePtr != IntPtr.Zero)
        {
            WindowNative.Destroy(_nativePtr);
            _nativePtr = IntPtr.Zero;
        }
    }

    public bool IsWindow() => WindowNative.IsWindow(_nativePtr);
}
